thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/api/detection/AnomalyDetectionResource.java

```diff
@@ -0,0 +1,756 @@
+        TimeSeriesCacheBuilder.getInstance(), AnomaliesCacheBuilder.getInstance());
+    this.detectionValidator = new DetectionConfigValidator(this.provider);
+    this.metricConfigValidator = new MetricConfigValidator();
+    this.datasetConfigValidator = new DatasetConfigValidator();
+
+    // Read template from disk
+    this.yaml = new Yaml();
+  }
+
+  /**
+   * Run an online anomaly detection service synchronously. It will run anomaly detection using
+   * default configs for detection, metric, dataset
+   *
+   * @param start     detection window start time
+   * @param end       detection window end time
+   * @param payload   payload in request including online data
+   * @param principal user who sent this request. It's used to separate different config names
+   * @return a message containing the detected anomalies and the detection config used
+   */
+  @POST
+  @Path("/")
+  @Produces(MediaType.APPLICATION_JSON)
+  @Consumes(MediaType.APPLICATION_JSON)
+  @ApiOperation("Request an anomaly detection online task")
+  public Response onlineApi(
+          @QueryParam("start") long start,
+          @QueryParam("end") long end,
+          @ApiParam("jsonPayload") String payload,
+          @Auth ThirdEyePrincipal principal) {
+    DatasetConfigDTO datasetConfigDTO = null;
+    MetricConfigDTO metricConfigDTO = null;
+    DetectionConfigDTO detectionConfigDTO = null;
+    TaskDTO taskDTO = null;
+    List<AnomalySummary> anomalies = null;
+    Response.Status responseStatus;
+    Map<String, String> responseMessage = new HashMap<>();
+    ObjectMapper objectMapper = new ObjectMapper();
+    // Use username to separate different requests. One user can only send one request at a time
+    String nameSuffix = "_" + principal.getName();
+
+    try {
+      if (payload.getBytes().length > MAX_ONLINE_PAYLOAD_SIZE) {
+        responseStatus = Response.Status.BAD_REQUEST;
+        responseMessage.put("message", "Payload too large");
+        return Response.status(responseStatus).entity(responseMessage).build();
+      }
+
+      JsonNode payloadNode = objectMapper.readTree(payload);
+
+      if (!validateOnlineRequestPayload(payloadNode)) {
+        responseStatus = Response.Status.BAD_REQUEST;
+        responseMessage.put("message", "Invalid request payload");
+        return Response.status(responseStatus).entity(responseMessage).build();
+      }
+
+      // Preprocess: remove existing entities generated by the previous interrupted request
+      cleanExistingOnlineTask(nameSuffix);
+
+      // Create & save dataset
+      datasetConfigDTO = generateDatasetConfig(payloadNode, nameSuffix);
+
+      // Create & save metric along with online data
+      metricConfigDTO = generateMetricConfig(payloadNode, nameSuffix);
+
+      // Create & save detection
+      detectionConfigDTO =
+          generateDetectionConfig(payloadNode, nameSuffix, datasetConfigDTO, metricConfigDTO, start,
+              end);
```

Comment:
I'm worried about creating a temporary detection config and later cleans it up. These detection configs can get leaked into the UI where users can see and change them while the task is running. 

Since we're already persisting in the task DTO, the information needed to create a detection config can actually just be persisted within the task info. Then an OnlineDetectionTaskRunner can pick it up and create the detectionConfigDTO in run time. No need to persist the DetectionConfig.

Reply:
Thanks for the suggestion! Now the online detection config is saved into task info instead of DB. Please check this [commit](https://github.com/apache/incubator-pinot/pull/5769/commits/3c5f97d2326d5d77b3c13203c1816e6c49da23fa).